File: java/GeigerApiPluginConnector/src/test/java/mocks/MessageCollector.java
package mocks;

import eu.cybergeiger.api.GeigerApi;
import eu.cybergeiger.api.message.Message;
import eu.cybergeiger.api.message.MessageType;
import eu.cybergeiger.api.plugin.PluginListener;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * A PluginListener that collects all messages it receives.
 */
public class MessageCollector implements PluginListener {
  public static long DEFAULT_TIMEOUT = 30000;

  final ArrayList<Message> messages = new ArrayList<>();
  private final Set<CountDownLatch> listeners = new HashSet<>();

  public MessageCollector(GeigerApi api){
    api.registerListener(MessageType.getAllTypes(), this);
  }

  public List<Message> getMessages() {
    return messages;
  }

  @Override
  public void pluginEvent(Message msg) {
    synchronized (messages) {
      synchronized (listeners) {
        messages.add(msg);
        for (CountDownLatch listener : listeners)
          listener.countDown();
      }
    }
  }

  public void awaitCount(int count) throws TimeoutException, InterruptedException {
    awaitCount(count, DEFAULT_TIMEOUT);
  }

  public void awaitCount(int count, long msTimeout) throws TimeoutException, InterruptedException {
    int remaining = count - messages.size();
    if (remaining <= 0) return;
    CountDownLatch listener = new CountDownLatch(remaining);
    synchronized (listeners) {
      listeners.add(listener);
    }
    try {
      if (!listener.await(msTimeout, TimeUnit.MILLISECONDS))
        throw new TimeoutException("Did not receive enough messages before timeout.");
    } finally {
      synchronized (listeners) {
        listeners.remove(listener);
      }
    }
  }

  public Message awaitMessage(int index) throws TimeoutException, InterruptedException {
    return awaitMessage(index, DEFAULT_TIMEOUT);
  }

  public Message awaitMessage(int index, long msTimeout) throws TimeoutException, InterruptedException {
    awaitCount(index+1, msTimeout);
    return getMessages().get(index);
  }
}

File: java/GeigerApiPluginConnector/src/test/dart/utils/message_collector.dart
import 'dart:async';

import 'package:geiger_api/geiger_api.dart';

class _ConditionalCompleter<TResult> {
  final bool Function(List<Message>) canComplete;
  final TResult Function(List<Message>) createResult;

  final _completer = Completer<TResult>();

  Future<TResult> get future => _completer.future;

  _ConditionalCompleter(this.canComplete, this.createResult);

  bool complete(List<Message> events) {
    if (!canComplete(events)) return false;
    _completer.complete(createResult(events));
    return true;
  }
}

class MessageCollector implements PluginListener {
  List<Message> messages = [];
  List<_ConditionalCompleter> _completers = [];

  MessageCollector(GeigerApi api) {
    api.registerListener(MessageType.values, this);
  }

  @override
  void pluginEvent(GeigerUrl? url, Message msg) {
    messages.add(msg);
    _completers = _completers
        .where((completer) => !completer.complete(messages))
        .toList();
  }

  Future awaitCount(int count,
      [Duration timeLimit = const Duration(seconds: 5)]) {
    canComplete(List messages) => messages.length >= count;
    if (canComplete(messages)) {
      return Future.value();
    }
    final completer = _ConditionalCompleter(canComplete, (_) => null);
    _completers.add(completer);
    return completer.future.then((value) {
      _completers.remove(completer);
      return value;
    }, onError: (error) {
      _completers.remove(completer);
      throw error;
    }).timeout(timeLimit, onTimeout: () {
      _completers.remove(completer);
      throw TimeoutException(
          'Did not receive enough messages before timeout.', timeLimit);
    });
  }

  Future<Message> awaitMessage(int index,
      [Duration timeLimit = const Duration(seconds: 5)]) async {
    await awaitCount(index + 1, timeLimit);
    return messages[index];
  }
}
